engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java

```diff
@@ -59,13 +63,16 @@ boolean isArrayElementReadable(long index) {
   }
 
   @ExportMessage
-  public Object readArrayElement(long index, @CachedLibrary(limit = "3") InteropLibrary interop)
+  public Object readArrayElement(
+      long index,
+      @Cached("create()") BranchProfile arrayIndexHasHappened,
+      @CachedLibrary(limit = "3") InteropLibrary interop)
       throws UnsupportedMessageException, InvalidArrayIndexException {
-    if (index >= length || index < 0) {
-      throw InvalidArrayIndexException.create(index);
-    }
-
     try {
+      if (index >= length || index < 0) {
+        arrayIndexHasHappened.enter();
```

Comment:
This profile gets me from 20us to 16us on my computer.

Reply:
That makes sense.  Why is it pulled into the try though?